### PR TITLE
Sanity checking libevent is initialized

### PR DESCRIPTION
**Pull request by alyssawilk**
We've had a couple of instances of internal Envoy with custom main or custom tests failing to initialize things properly and causing hard to debug timeouts / tsan warnings.

Adding RELEASE_ASSERTs to hopefully catch this earlier and in a more obvious way.

*Risk Level*: Low
*Testing*: n/a
*Release Notes*: n/a

### CODE REVIEW

**Comment by alyssawilk** [comment]
Sure!  I'd only used ASSERT I figured since it is way more likely to affect Googlers (using our internal version of gtest) and this would minimize the impact on everyone else.

**Comment by ggreenway** [comment]
I think I'm in the minority, but I prefer all ASSERTs to be RELEASE_ASSERT unless it has a large performance impact (like if a bunch of code had to be run to calculate the thing that is being asserted).  I'd rather have it abort than run in an unexpected/unsupported state.

File: source/common/event/dispatcher_impl.cc [hunk 1]
```diff
@@ -26,13 +26,18 @@ namespace Envoy {
...
+  RELEASE_ASSERT(Libevent::Global::initialized());
+}
 
 DispatcherImpl::DispatcherImpl(Buffer::WatermarkFactoryPtr&& factory)
     : buffer_factory_(std::move(factory)), base_(event_base_new()),
       deferred_delete_timer_(createTimer([this]() -> void { clearDeferredDeleteList(); })),
       post_timer_(createTimer([this]() -> void { runPostCallbacks(); })),
-      current_to_delete_(&to_delete_1_) {}
+      current_to_delete_(&to_delete_1_) {
+  ASSERT(Libevent::Global::initialized());
```
Comment (ggreenway) [inline source/common/event/dispatcher_impl.cc]: Oh, this one should probably also be RELEASE_ASSERT

File: source/common/event/libevent.h [hunk 2]
```diff
@@ -31,6 +31,11 @@ namespace Libevent {
  */
 class Global {
 public:
+  // True if initialized() has been called.
+  static bool INITIALIZED;
```
Comment (htuch) [inline source/common/event/libevent.h]: I think this should be lower case, this isn't a constant?

Comment (mattklein123) [reply source/common/event/libevent.h]: nit: This variable can be private.

Comment (alyssawilk) [reply source/common/event/libevent.h]: Oh, is the CAPS only for contants?  I thought it was for all static.
Ignoring the naming conflict, static bool intitialized; is weird and static bool initialized_ isn't correct.

File: source/common/event/dispatcher_impl.cc (hunk 1 above)
Comment (alyssawilk) [reply source/common/event/dispatcher_impl.cc]: What are you talking about, I already changed that file! ... *facepalm*  :-P

File: source/common/event/libevent.h [hunk 3]
```diff
@@ -31,10 +31,16 @@ namespace Libevent {
...
+  static bool initialized() { return is_initialized; }
+
   /**
    * Initialize the library globally.
    */
   static void initialize();
+
+private:
+  // True if initialized() has been called.
+  static bool is_initialized;
```
Comment (ggreenway) [inline source/common/event/libevent.h]: I think is_initialized_ or initialized_ is the correct name, even though it is static:

https://google.github.io/styleguide/cppguide.html#Variable_Names section "Class Data Members", and I didn't see an override in STYLE.md

Comment (mattklein123) [reply source/common/event/libevent.h]: +1

**Pull request merged**